### DISCUSSION TITLE
`valet run` to use correct php version automatically for php cli calls

### DISCUSSION THREAD

**Discussion by tpetry** [laravel/valet#1211] [Ideas]
I've proposed [on twitter](https://twitter.com/tobias_petry/status/1504471584785936384?s=20&t=ETEqRRzgvNBRuSwxSDa5jA) an idea on how to improve the valet isolation feature for php cli commands. Changing the global php version to the one from the current valet is problematic if you are working on multiple valet sites in parallel with different versions. You would have  to permanently change the global php version.

How about a new command like `valet run [args]` which is using the php version of the current valet site (the cli is within) to run the command? A call like `valet run composer update` will work like this:

```
# within ~/site1 [no-isolation]
php composer update

# within ~/site2 [8.0]
php80 composer update

# within ~/site3 [8.1]
php81 composer update
```

So instead of constantly changing the global php version when working on multiple projects in parallel, the `valet run` command will use the valet information about the current directory to run commands. No more use of `valet use` is ever needed!

**Comment by JacobDelcroix** [comment]
As discuss on twitter we definitly need that feature to fully isolate our projects. I’m sure many people will try to run artisan or composer command without thinking to change the global php version before. Hope this idea will be accepted! If i can help you to develop that, I will for sure.

**Comment by NasirNobin** [comment]
I've been trying to solve this same issue. I kinda have a `zsh` plugin ready that would automatically swap the PHP binary based on the project directory you are in.

I know the Valet team will release something officially. In but meantime, if anyone is interested Please give my solution a try.  This is very early, feedbacks are welcome!

https://github.com/NasirNobin/zsh-valet 

<details>
<summary>Here's a quick video demonstration </summary>

https://user-images.githubusercontent.com/13833460/158659897-f6376d8d-8dfa-4e2d-a82a-82bb0aafb009.mp4
</details>

**Comment by adrianspacely** [comment]
I left a comment about this yesterday too.

Forge has `php7.4`, Valet could do the same. Otherwise `valet php ...` might be good if you don't want to remember which PHP version the site uses.

https://github.com/laravel/valet/issues/1206#issuecomment-1070325542

**Comment by mattstauffer** [comment]
I like this idea and I'm more than happy to welcome a pull request. In the meantime, I'll make plans to work on it eventually but don't have an immediate timeline.

I think the idea of `valet run` is really interesting way of addressing this problem, because it drastically simplifies the implementation costs _and_ can easily be aliased to something like `vhp` or whatever an individual would want to use to use "Valet's current version of PHP".

**Comment by NasirNobin** [comment]
Submitted a PR by adding this functionality here https://github.com/laravel/valet/pull/1216 🙌

**Comment by ChrisHardie** [comment]
For anyone using WordPress and WP CLI with valet, this worked for me to make sure WP CLI uses the right PHP version (in this example, to use the `shell` subcommand):

```
$ valet php /opt/homebrew/bin/wp shell
```

May need to be updated with whatever the path is to your wp binary, and ripe for setting up as a command line shell alias.
